web/modules/custom/joinup_discussion/src/EventSubscriber/SubscribedDiscussionSubscriber.php

```diff
@@ -76,56 +99,159 @@ public static function getSubscribedEvents(): array {
+
+    // Exclude the current user from the list.
+    $current_user = $this->getCurrentUser();
+    return array_filter($subscribers, function (AccountInterface $subscriber) use ($current_user) {
+      return $subscriber->id() != $current_user->id();
+    });
   }
 
   /**
-   * Builds the message arguments.
+   * Returns the message arguments.
+   *
+   * @param \Drupal\node\NodeInterface $discussion
+   *   The discussion for which to return the arguments.
+   *
+   * @return array
+   *   An associative array of message arguments, keyed by argument ID.
+   *
+   * @throws \Exception
+   *   Thrown if one of the arguments could not be generated.
    */
-  protected function getArguments(): array {
-    return [
-      '@entity:title' => $this->discussion->label(),
-      '@group:label' => $this->group->label(),
-      '@group:bundle' => $this->group->bundle(),
-      '@entity:url' => $this->discussion->toUrl('canonical', [
-        'absolute' => TRUE,
-      ])->toString(),
-    ];
+  protected function getArguments(NodeInterface $discussion) : array {
+    $arguments = [];
+
+    $arguments['@entity:title'] = $discussion->label();
+    try {
+      $entity_url = $discussion->toUrl('canonical', ['absolute' => TRUE])->toString();
+    }
+    catch (EntityMalformedException | UndefinedLinkTemplateException $e) {
+      // Retrieval of the URL might fail if the link template is not defined, or
+      // if the entity doesn't have an ID.
+      throw new \Exception('Could not retrieve URL for discussion.', NULL, $e);
+    }
+    $arguments['@entity:url'] = $entity_url;
+
+    $actor = $this->getCurrentUser();
+    $actor_first_name = !empty($actor->get('field_user_first_name')->value) ? $actor->get('field_user_first_name')->value : '';
+    $actor_family_name = !empty($actor->get('field_user_family_name')->value) ? $actor->get('field_user_family_name')->value : '';
+
+    if ($actor->hasRole('moderator')) {
+      $arguments['@actor:full_name'] = 'The Joinup Support Team';
+    }
+    else {
+      $arguments['@actor:full_name'] = empty($actor->get('full_name')->value) ? $actor_first_name . ' ' . $actor_family_name : $actor->get('full_name')->value;
+    }
+
+    $group = $this->getDiscussionGroup($discussion);
+    if ($group) {
+      $arguments['@group:label'] = $group->label();
+      $arguments['@group:bundle'] = $group->bundle();
+    }
+
+    return $arguments;
+  }
+
+  /**
+   * Returns the fully loaded User entity for the current user.
+   *
+   * @return \Drupal\user\UserInterface
+   *   The current user.
+   */
+  protected function getCurrentUser(): UserInterface {
+    try {
+      /** @var \Drupal\user\UserInterface $user */
+      $user = $this->entityTypeManager->getStorage('user')->load($this->currentUser->id());
+      return $user;
+    }
+    catch (InvalidPluginDefinitionException $e) {
+      // The storage for the 'user' entity type should exist. Ignore this.
```

Comment:
True, but it would be nice to return an explicit `NULL`. At least the IDE would not complain :)

Reply:
Implementing this I found that this method shouldn't return NULL because then the problem just cascades up the call stack and the IDE starts complaining in other methods. We positively need to be certain that this method returns a user. I found a creative solution for it that satisfies PHPStorm :D